File: typescript/700/700.search-in-a-binary-search-tree.ts
/*
 * @lc app=leetcode id=700 lang=typescript
 *
 * [700] Search in a Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function searchBST(root: TreeNode | null, val: number): TreeNode | null {
    return search(root, val);
};

function search(treeNode: TreeNode | null, val: number): TreeNode | null {
    if (!treeNode) {
        return null;
    }

    if (treeNode.val === val) {
        return treeNode;
    }

    if (treeNode.val > val) {
        return search(treeNode.left, val);
    } else  {
        return search(treeNode.right, val);
    }
}
// @lc code=end

File: rust/src/bin/700/700.search-in-a-binary-search-tree.rs
/*
 * @lc app=leetcode id=700 lang=rust
 *
 * [700] Search in a Binary Search Tree
 */

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }

// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn search_bst(root: Option<Rc<RefCell<TreeNode>>>, val: i32) -> Option<Rc<RefCell<TreeNode>>> {
        Self::search(root, val)
    }

    fn search(tree: Option<Rc<RefCell<TreeNode>>>, val: i32) -> Option<Rc<RefCell<TreeNode>>> {
        match tree {
            None => None,
            Some(treeNode) => {
                let current_val = treeNode.borrow().val;
                if current_val > val {
                    Self::search(treeNode.borrow().left.clone(), val)
                } else if current_val < val {
                    Self::search(treeNode.borrow().right.clone(), val)
                } else{
                    Some(treeNode)
                }
            }
        }
    }
}
// @lc code=end
